File: react_dashboard/src/engine/scheduler.js
/* ═══════════════════════════════════════════════════════════════════
 * scheduler.js  –  Event-Driven CPU Scheduling Engine (JS)
 * Instruments state changes and logs for time-travel playback.
 * ═══════════════════════════════════════════════════════════════════ */

export const ALGO = { FCFS: 0, SJF: 1, SRTF: 2, RR: 3, PRIORITY: 4 };

export const ALGO_NAMES = {
  [ALGO.FCFS]:     'FCFS',
  [ALGO.SJF]:      'SJF (AI)',
  [ALGO.SRTF]:     'SRTF (AI)',
  [ALGO.RR]:       'Round Robin',
  [ALGO.PRIORITY]: 'Priority',
};

export const STATE = {
  NEW:     'NEW',
  READY:   'READY',
  RUNNING: 'RUNNING',
  WAITING: 'WAITING',
  DONE:    'DONE',
};

export const PROCESS_COLORS = [
  { from: '#7c3aed', to: '#a78bfa', hex: '#8b5cf6', name: 'violet' },
  { from: '#0e7490', to: '#22d3ee', hex: '#06b6d4', name: 'cyan'   },
  { from: '#059669', to: '#34d399', hex: '#10b981', name: 'emerald'},
  { from: '#be185d', to: '#f472b6', hex: '#ec4899', name: 'pink'   },
  { from: '#b45309', to: '#fbbf24', hex: '#f59e0b', name: 'amber'  },
  { from: '#1d4ed8', to: '#60a5fa', hex: '#3b82f6', name: 'blue'   },
  { from: '#0f766e', to: '#2dd4bf', hex: '#14b8a6', name: 'teal'   },
  { from: '#a21caf', to: '#e879f9', hex: '#d946ef', name: 'fuchsia'},
  { from: '#9a3412', to: '#fb923c', hex: '#f97316', name: 'orange' },
  { from: '#166534', to: '#86efac', hex: '#22c55e', name: 'green'  },
];

// ── Min-heap event queue ───────────────────────────────────────────
class MinHeap {
  constructor() { this.h = []; }
  push(e) {
    this.h.push(e);
    let i = this.h.length - 1;
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (this._lt(this.h[i], this.h[p])) { [this.h[i], this.h[p]] = [this.h[p], this.h[i]]; i = p; }
      else break;
    }
  }
  pop() {
    const top = this.h[0];
    const last = this.h.pop();
    if (this.h.length) { this.h[0] = last; this._down(0); }
    return top;
  }
  get size() { return this.h.length; }
  _lt(a, b) { return a.time !== b.time ? a.time < b.time : a.type < b.type; }
  _down(i) {
    const n = this.h.length;
    while (true) {
      let s = i, l = 2*i+1, r = 2*i+2;
      if (l < n && this._lt(this.h[l], this.h[s])) s = l;
      if (r < n && this._lt(this.h[r], this.h[s])) s = r;
      if (s === i) break;
      [this.h[s], this.h[i]] = [this.h[i], this.h[s]];
      i = s;
    }
  }
}

const EVT = { ARRIVAL: 0, CPU_END: 1, IO_END: 2, QUANTUM: 3 };

// ── AI predictor ───────────────────────────────────────────────────
function predict(p, actual) {
  p.predicted = p.alpha * actual + (1 - p.alpha) * p.predicted;
  return p.predicted;
}

// ── Pick next from ready queue ─────────────────────────────────────
function pickNext(procs, ready, algo) {
  if (!ready.length) return -1;
  let best = 0;
  for (let i = 1; i < ready.length; i++) {
    const b = procs[ready[best]], c = procs[ready[i]];
    switch (algo) {
      case ALGO.FCFS:     if (c.arrivalTime < b.arrivalTime) best = i; break;
      case ALGO.SJF:
      case ALGO.SRTF:     if (c.predicted < b.predicted) best = i; break;
      case ALGO.PRIORITY: if (c.priority < b.priority) best = i; break;
      case ALGO.RR:       best = 0; i = ready.length; break;
    }
  }
  const pid = ready[best];
  ready.splice(best, 1);
  return pid;
}

// ── Gantt push ─────────────────────────────────────────────────────
function ganttPush(gantt, start, end, pid) {
  if (start >= end) return;
  const last = gantt[gantt.length - 1];
  if (last && last.pid === pid && last.end === start) { last.end = end; return; }
  gantt.push({ start, end, pid });
}

// ═══════════════════════════════════════════════════════════════════
// Main simulation
// ═══════════════════════════════════════════════════════════════════
export function runSimulation(inputProcs, algo, quantum = 2) {
  const stateChanges = [];  // {time, pid, state, seq}
  let seq = 0;
  const logs = [];          // {time, message, type, id}
  let logId = 0;

  function recordState(time, pid, state) {
    stateChanges.push({ time, pid, state, seq: seq++ });
  }
  function log(time, message, type = 'info') {
    logs.push({ time, message, type, id: logId++ });
  }

  // Deep clone + init runtime fields
  const procs = inputProcs.map(p => ({
    pid:           p.pid,
    priority:      p.priority ?? 1,
    arrivalTime:   p.arrivalTime,
    bursts:        p.bursts.map(b => ({ cpu: b.cpu, io: b.io ?? 0 })),
    state:         STATE.NEW,
    burstIdx:      0,
    remainingCpu:  p.bursts[0].cpu,
    finishTime:    0,
    waitingTime:   0,
    turnaroundTime: 0,
    responseTime:  -1,
    firstRun:      false,
    predicted:     p.bursts[0].cpu,
    alpha:         p.alpha ?? 0.5,
  }));

  const eq    = new MinHeap();
  const ready = [];
  const gantt = [];

  let currentPid = -1;
  let busyStart  = 0;
  let totalBusy  = 0;
  let qEnd       = 0;

  // Seed arrivals
  procs.forEach(p => {
    eq.push({ time: p.arrivalTime, type: EVT.ARRIVAL, pid: p.pid });
  });

  const startProc = (pid, now) => {
    currentPid = pid;
    procs[pid].state = STATE.RUNNING;
    if (!procs[pid].firstRun) {
      procs[pid].firstRun = true;
      procs[pid].responseTime = now - procs[pid].arrivalTime;
    }
    recordState(now, pid, STATE.RUNNING);
    log(now, `P${pid} dispatched to CPU (${procs[pid].remainingCpu}ms remaining)`, 'cpu');
    busyStart = now;
    const burst = procs[pid].remainingCpu;
    if (algo === ALGO.RR) {
      const run = Math.min(burst, quantum);
      qEnd = now + run;
      eq.push({ time: qEnd, type: EVT.QUANTUM, pid });
    } else {
      eq.push({ time: now + burst, type: EVT.CPU_END, pid });
    }
  };

  const finishBurst = (pid, now) => {
    const bi = procs[pid].burstIdx;
    const predBefore = procs[pid].predicted;
    predict(procs[pid], procs[pid].bursts[bi].cpu);
    const io = procs[pid].bursts[bi].io;

    if ((algo === ALGO.SJF || algo === ALGO.SRTF) && bi + 1 < procs[pid].bursts.length) {
      log(now, `AI predicted next burst for P${pid}: ${procs[pid].predicted.toFixed(1)}ms (was ${predBefore.toFixed(1)}ms)`, 'ai');
    }

    if (io > 0 && bi + 1 < procs[pid].bursts.length) {
      procs[pid].state = STATE.WAITING;
      recordState(now, pid, STATE.WAITING);
      log(now, `P${pid} entering I/O wait (${io}ms)`, 'io');
      eq.push({ time: now + io, type: EVT.IO_END, pid });
    } else {
      procs[pid].state = STATE.DONE;
      procs[pid].finishTime = now;
      procs[pid].turnaroundTime = now - procs[pid].arrivalTime;
      recordState(now, pid, STATE.DONE);
      log(now, `P${pid} completed — TAT: ${procs[pid].turnaroundTime}ms`, 'done');
    }
  };

  while (eq.size > 0) {
    const ev  = eq.pop();
    const now = ev.time;

    switch (ev.type) {

    case EVT.ARRIVAL: {
      procs[ev.pid].state = STATE.READY;
      ready.push(ev.pid);
      recordState(now, ev.pid, STATE.READY);
      log(now, `P${ev.pid} arrived (burst: ${procs[ev.pid].bursts[0].cpu}ms, priority: ${procs[ev.pid].priority})`, 'arrival');

      if (currentPid === -1) {
        const next = pickNext(procs, ready, algo);
        if (next !== -1) startProc(next, now);
      } else if (algo === ALGO.SRTF) {
        const np = procs[ev.pid], cp = procs[currentPid];
        if (np.predicted < cp.remainingCpu) {
          const elapsed = now - busyStart;
          totalBusy += elapsed;
          ganttPush(gantt, busyStart, now, currentPid);
          cp.remainingCpu -= elapsed;
          cp.state = STATE.READY;
          ready.push(currentPid);
          recordState(now, currentPid, STATE.READY);
          log(now, `P${ev.pid} preempts P${currentPid} (SRTF)`, 'preempt');
          currentPid = -1;
          ready.splice(ready.indexOf(ev.pid), 1);
          startProc(ev.pid, now);
        }
      }
      break;
    }

    case EVT.CPU_END: {
      if (ev.pid !== currentPid) break;
      const elapsed = now - busyStart;
      totalBusy += elapsed;
      ganttPush(gantt, busyStart, now, ev.pid);
      finishBurst(ev.pid, now);
      currentPid = -1;
      const next = pickNext(procs, ready, algo);
      if (next !== -1) startProc(next, now);
      break;
    }

    case EVT.IO_END: {
      const pid = ev.pid;
      procs[pid].burstIdx++;
      const bi = procs[pid].burstIdx;
      procs[pid].remainingCpu = procs[pid].bursts[bi].cpu;
      procs[pid].predicted    = procs[pid].remainingCpu; // re-seed
      procs[pid].state = STATE.READY;
      ready.push(pid);
      recordState(now, pid, STATE.READY);
      log(now, `P${pid} I/O complete, returned to Ready queue`, 'io');

      if (currentPid === -1) {
        const next = pickNext(procs, ready, algo);
        if (next !== -1) startProc(next, now);
      } else if (algo === ALGO.SRTF) {
        const np = procs[pid], cp = procs[currentPid];
        if (np.predicted < cp.remainingCpu) {
          const elapsed = now - busyStart;
          totalBusy += elapsed;
          ganttPush(gantt, busyStart, now, currentPid);
          cp.remainingCpu -= elapsed;
          cp.state = STATE.READY;
          ready.push(currentPid);
          recordState(now, currentPid, STATE.READY);
          log(now, `P${pid} preempts P${currentPid} after I/O (SRTF)`, 'preempt');
          currentPid = -1;
          ready.splice(ready.indexOf(pid), 1);
          startProc(pid, now);
        }
      }
      break;
    }

    case EVT.QUANTUM: {
      if (ev.pid !== currentPid || now !== qEnd) break;
      const elapsed = now - busyStart;
      totalBusy += elapsed;
      ganttPush(gantt, busyStart, now, ev.pid);
      procs[ev.pid].remainingCpu -= elapsed;

      if (procs[ev.pid].remainingCpu <= 0) {
        finishBurst(ev.pid, now);
      } else {
        procs[ev.pid].state = STATE.READY;
        ready.push(ev.pid);
        recordState(now, ev.pid, STATE.READY);
        log(now, `P${ev.pid} quantum expired (${procs[ev.pid].remainingCpu}ms remaining)`, 'preempt');
      }
      currentPid = -1;
      const next = pickNext(procs, ready, algo);
      if (next !== -1) startProc(next, now);
      break;
    }
    }
  }

  // Compute final metrics
  let totalTime = 0, sumWt = 0, sumTat = 0, sumRt = 0, finished = 0;
  procs.forEach(p => {
    if (p.state === STATE.DONE) {
      const totalCpu = p.bursts.reduce((s, b) => s + b.cpu, 0);
      p.waitingTime  = Math.max(0, p.turnaroundTime - totalCpu);
      sumWt  += p.waitingTime;
      sumTat += p.turnaroundTime;
      sumRt  += Math.max(0, p.responseTime);
      finished++;
      if (p.finishTime > totalTime) totalTime = p.finishTime;
    }
  });

  log(totalTime, `Simulation complete. ${finished} processes. Total time: ${totalTime}ms`, 'done');

  return {
    processes:          procs,
    gantt,
    stateChanges,
    logs,
    totalTime,
    avgWaitingTime:     finished ? sumWt  / finished : 0,
    avgTurnaroundTime:  finished ? sumTat / finished : 0,
    avgResponseTime:    finished ? sumRt  / finished : 0,
    cpuUtilization:     totalTime ? totalBusy / totalTime * 100 : 0,
    throughput:         totalTime ? finished / totalTime * 1000  : 0,
  };
}

// ── Time-travel: get process states at a given time T ──────────────
export function getStatesAtTime(procs, stateChanges, T) {
  return procs.map(p => {
    const changes = stateChanges
      .filter(c => c.pid === p.pid && c.time <= T)
      .sort((a, b) => b.time !== a.time ? b.time - a.time : b.seq - a.seq);
    return { ...p, currentState: changes[0]?.state ?? STATE.NEW };
  });
}

// ── Demo processes ─────────────────────────────────────────────────
export const DEMO_PROCESSES = [
  { pid: 0, priority: 2, arrivalTime: 0,  bursts: [{ cpu: 6, io: 4 }, { cpu: 3, io: 0 }], alpha: 0.5 },
  { pid: 1, priority: 1, arrivalTime: 2,  bursts: [{ cpu: 4, io: 0 }],                    alpha: 0.5 },
  { pid: 2, priority: 3, arrivalTime: 4,  bursts: [{ cpu: 8, io: 2 }, { cpu: 1, io: 0 }], alpha: 0.5 },
  { pid: 3, priority: 1, arrivalTime: 6,  bursts: [{ cpu: 2, io: 0 }],                    alpha: 0.5 },
  { pid: 4, priority: 2, arrivalTime: 8,  bursts: [{ cpu: 5, io: 3 }, { cpu: 2, io: 0 }], alpha: 0.5 },
];

export function randInt(lo, hi) { return Math.floor(Math.random() * (hi - lo + 1)) + lo; }

export function generateRandomProcess(pid, arrivalTime) {
  const burstCount = randInt(1, 3);
  const bursts = [];
  for (let i = 0; i < burstCount; i++) {
    bursts.push({ cpu: randInt(2, 12), io: i < burstCount - 1 ? randInt(1, 5) : 0 });
  }
  return { pid, priority: randInt(1, 5), arrivalTime: arrivalTime ?? randInt(0, 10), bursts, alpha: 0.5 };
}

File: c_engine/scheduler.c
#include "scheduler.h"
#include <string.h>
#include <stdlib.h>

/* ═══════════════════════════════════════════════════════════════════════════
 *  Internal helpers
 * ═══════════════════════════════════════════════════════════════════════════*/

/* Min-heap event queue */
static Event   eq[MAX_EVENTS];
static int     eq_size = 0;

static void eq_push(Event e) {
    int i = eq_size++;
    eq[i] = e;
    /* bubble up */
    while (i > 0) {
        int parent = (i - 1) / 2;
        if (eq[parent].time > eq[i].time ||
           (eq[parent].time == eq[i].time && eq[parent].type > eq[i].type)) {
            Event tmp = eq[parent]; eq[parent] = eq[i]; eq[i] = tmp;
            i = parent;
        } else break;
    }
}

static Event eq_pop(void) {
    Event top = eq[0];
    eq[0] = eq[--eq_size];
    /* bubble down */
    int i = 0;
    while (1) {
        int l = 2*i+1, r = 2*i+2, smallest = i;
        if (l < eq_size && (eq[l].time < eq[smallest].time ||
           (eq[l].time == eq[smallest].time && eq[l].type < eq[smallest].type)))
            smallest = l;
        if (r < eq_size && (eq[r].time < eq[smallest].time ||
           (eq[r].time == eq[smallest].time && eq[r].type < eq[smallest].type)))
            smallest = r;
        if (smallest == i) break;
        Event tmp = eq[smallest]; eq[smallest] = eq[i]; eq[i] = tmp;
        i = smallest;
    }
    return top;
}

/* Ready queue (simple array, re-sorted each time we pick next process) */
static int ready[MAX_PROCESSES];
static int ready_size = 0;

static void ready_push(int pid) {
    ready[ready_size++] = pid;
}

static void ready_remove(int pid) {
    for (int i = 0; i < ready_size; i++) {
        if (ready[i] == pid) {
            ready[i] = ready[--ready_size];
            return;
        }
    }
}

/* ── AI predictor ───────────────────────────────────────────────────────── */
double predict_next_burst(PCB *p, int actual_burst) {
    p->predicted_burst = p->alpha * actual_burst + (1.0 - p->alpha) * p->predicted_burst;
    return p->predicted_burst;
}

/* ── Pick next process from ready queue based on algorithm ─────────────── */
static int pick_next(PCB *procs, Algorithm algo) {
    if (ready_size == 0) return -1;
    int best = 0;
    for (int i = 1; i < ready_size; i++) {
        PCB *b = &procs[ready[best]];
        PCB *c = &procs[ready[i]];
        switch (algo) {
            case ALGO_FCFS:
                if (c->arrival_time < b->arrival_time) best = i;
                break;
            case ALGO_SJF:
            case ALGO_SRTF:
                if (c->predicted_burst < b->predicted_burst) best = i;
                break;
            case ALGO_PRIORITY:
                if (c->priority < b->priority) best = i;
                break;
            case ALGO_RR:
                /* FIFO order – pick the one that arrived in ready queue first
                   We encode arrival order via a simple counter stored in arrival_time
                   for the ready queue (we use a separate field). For RR we just pick
                   index 0 (FIFO). */
                best = 0;
                i = ready_size; /* break loop */
                break;
        }
    }
    int pid = ready[best];
    /* remove from ready */
    ready[best] = ready[--ready_size];
    return pid;
}

/* ── Gantt helper ──────────────────────────────────────────────────────── */
static void gantt_push(SimResult *res, int start, int end, int pid) {
    if (start >= end) return;
    /* merge with last slice if same pid */
    if (res->gantt_count > 0) {
        GanttSlice *last = &res->gantt[res->gantt_count - 1];
        if (last->pid == pid && last->end == start) {
            last->end = end;
            return;
        }
    }
    if (res->gantt_count < MAX_EVENTS) {
        res->gantt[res->gantt_count++] = (GanttSlice){start, end, pid};
    }
}

/* ═══════════════════════════════════════════════════════════════════════════
 *  Main simulation
 * ═══════════════════════════════════════════════════════════════════════════*/
void simulate(PCB *procs, int n, Algorithm algo, int time_quantum, SimResult *result) {
    memset(result, 0, sizeof(*result));
    result->process_count = n;

    /* copy processes into result */
    for (int i = 0; i < n; i++) {
        result->processes[i] = procs[i];
        result->processes[i].state           = STATE_NEW;
        result->processes[i].current_burst_idx = 0;
        result->processes[i].remaining_cpu   = procs[i].bursts[0].cpu_burst;
        result->processes[i].finish_time     = 0;
        result->processes[i].waiting_time    = 0;
        result->processes[i].turnaround_time = 0;
        result->processes[i].response_time   = -1;
        result->processes[i].first_run       = 0;
        /* initialise predicted burst to first actual burst */
        result->processes[i].predicted_burst = procs[i].bursts[0].cpu_burst;
        if (result->processes[i].alpha == 0.0)
            result->processes[i].alpha = 0.5;
    }

    PCB *P = result->processes;

    /* reset event queue and ready queue */
    eq_size    = 0;
    ready_size = 0;

    /* seed arrival events */
    for (int i = 0; i < n; i++) {
        eq_push((Event){P[i].arrival_time, EVT_ARRIVAL, i});
        P[i].state = STATE_NEW;
    }

    int current_pid   = -1;  /* pid currently on CPU, -1 = idle */
    int cpu_free_at   = 0;
    int busy_start    = 0;
    int total_busy    = 0;

    /* For RR quantum tracking */
    int quantum_end   = 0;

    while (eq_size > 0) {
        Event ev = eq_pop();
        int   now = ev.time;

        switch (ev.type) {

        /* ── Process arrives ───────────────────────────────────────────── */
        case EVT_ARRIVAL: {
            int pid = ev.pid;
            P[pid].state = STATE_READY;
            ready_push(pid);

            /* If CPU is free, schedule immediately */
            if (current_pid == -1) {
                int next = pick_next(P, algo);
                if (next == -1) break;
                current_pid = next;
                P[current_pid].state = STATE_RUNNING;
                if (!P[current_pid].first_run) {
                    P[current_pid].first_run   = 1;
                    P[current_pid].response_time = now - P[current_pid].arrival_time;
                }
                busy_start = now;
                int burst = P[current_pid].remaining_cpu;
                if (algo == ALGO_RR) {
                    int run = (burst < time_quantum) ? burst : time_quantum;
                    quantum_end = now + run;
                    eq_push((Event){quantum_end, EVT_QUANTUM_EXPIRE, current_pid});
                } else {
                    eq_push((Event){now + burst, EVT_CPU_BURST_END, current_pid});
                }
            } else if (algo == ALGO_SRTF) {
                /* check if new process preempts current */
                if (P[pid].predicted_burst < P[current_pid].remaining_cpu) {
                    /* preempt */
                    int elapsed = now - busy_start;
                    total_busy += elapsed;
                    gantt_push(result, busy_start, now, current_pid);
                    P[current_pid].remaining_cpu -= elapsed;
                    P[current_pid].state = STATE_READY;
                    ready_push(current_pid);

                    current_pid = pid;
                    ready_remove(pid);
                    P[current_pid].state = STATE_RUNNING;
                    if (!P[current_pid].first_run) {
                        P[current_pid].first_run    = 1;
                        P[current_pid].response_time = now - P[current_pid].arrival_time;
                    }
                    busy_start  = now;
                    eq_push((Event){now + P[current_pid].remaining_cpu, EVT_CPU_BURST_END, current_pid});
                }
            }
            break;
        }

        /* ── CPU burst finished ────────────────────────────────────────── */
        case EVT_CPU_BURST_END: {
            int pid = ev.pid;
            if (pid != current_pid) break; /* stale event */

            int elapsed = now - busy_start;
            total_busy += elapsed;
            gantt_push(result, busy_start, now, pid);

            /* update prediction */
            predict_next_burst(&P[pid], P[pid].bursts[P[pid].current_burst_idx].cpu_burst);

            int bi = P[pid].current_burst_idx;
            int io  = P[pid].bursts[bi].io_burst;

            if (io > 0 && bi + 1 < P[pid].burst_count) {
                /* go to I/O */
                P[pid].state = STATE_WAITING;
                eq_push((Event){now + io, EVT_IO_END, pid});
            } else {
                /* process finished */
                P[pid].state          = STATE_DONE;
                P[pid].finish_time    = now;
                P[pid].turnaround_time = now - P[pid].arrival_time;
            }

            current_pid = -1;
            cpu_free_at = now;

            /* schedule next ready process */
            int next = pick_next(P, algo);
            if (next != -1) {
                current_pid = next;
                P[current_pid].state = STATE_RUNNING;
                if (!P[current_pid].first_run) {
                    P[current_pid].first_run    = 1;
                    P[current_pid].response_time = now - P[current_pid].arrival_time;
                }
                busy_start = now;
                int burst = P[current_pid].remaining_cpu;
                if (algo == ALGO_RR) {
                    int run = (burst < time_quantum) ? burst : time_quantum;
                    quantum_end = now + run;
                    eq_push((Event){quantum_end, EVT_QUANTUM_EXPIRE, current_pid});
                } else {
                    eq_push((Event){now + burst, EVT_CPU_BURST_END, current_pid});
                }
            }
            break;
        }

        /* ── I/O finished ──────────────────────────────────────────────── */
        case EVT_IO_END: {
            int pid = ev.pid;
            P[pid].current_burst_idx++;
            int bi = P[pid].current_burst_idx;
            P[pid].remaining_cpu = P[pid].bursts[bi].cpu_burst;
            P[pid].predicted_burst = P[pid].remaining_cpu; /* re-seed for new burst */
            P[pid].state = STATE_READY;
            ready_push(pid);

            if (current_pid == -1) {
                int next = pick_next(P, algo);
                if (next == -1) break;
                current_pid = next;
                P[current_pid].state = STATE_RUNNING;
                if (!P[current_pid].first_run) {
                    P[current_pid].first_run    = 1;
                    P[current_pid].response_time = now - P[current_pid].arrival_time;
                }
                busy_start = now;
                int burst = P[current_pid].remaining_cpu;
                if (algo == ALGO_RR) {
                    int run = (burst < time_quantum) ? burst : time_quantum;
                    quantum_end = now + run;
                    eq_push((Event){quantum_end, EVT_QUANTUM_EXPIRE, current_pid});
                } else {
                    eq_push((Event){now + burst, EVT_CPU_BURST_END, current_pid});
                }
            } else if (algo == ALGO_SRTF) {
                if (P[pid].predicted_burst < P[current_pid].remaining_cpu) {
                    int elapsed = now - busy_start;
                    total_busy += elapsed;
                    gantt_push(result, busy_start, now, current_pid);
                    P[current_pid].remaining_cpu -= elapsed;
                    P[current_pid].state = STATE_READY;
                    ready_push(current_pid);

                    current_pid = pid;
                    ready_remove(pid);
                    P[current_pid].state = STATE_RUNNING;
                    if (!P[current_pid].first_run) {
                        P[current_pid].first_run    = 1;
                        P[current_pid].response_time = now - P[current_pid].arrival_time;
                    }
                    busy_start = now;
                    eq_push((Event){now + P[current_pid].remaining_cpu, EVT_CPU_BURST_END, current_pid});
                }
            }
            break;
        }

        /* ── Round Robin quantum expired ───────────────────────────────── */
        case EVT_QUANTUM_EXPIRE: {
            int pid = ev.pid;
            if (pid != current_pid) break; /* stale */
            if (now != quantum_end)   break; /* stale */

            int elapsed = now - busy_start;
            total_busy += elapsed;
            gantt_push(result, busy_start, now, pid);

            P[pid].remaining_cpu -= elapsed;

            if (P[pid].remaining_cpu <= 0) {
                /* burst finished exactly on quantum boundary */
                predict_next_burst(&P[pid], P[pid].bursts[P[pid].current_burst_idx].cpu_burst);
                int bi = P[pid].current_burst_idx;
                int io  = P[pid].bursts[bi].io_burst;
                if (io > 0 && bi + 1 < P[pid].burst_count) {
                    P[pid].state = STATE_WAITING;
                    eq_push((Event){now + io, EVT_IO_END, pid});
                } else {
                    P[pid].state          = STATE_DONE;
                    P[pid].finish_time    = now;
                    P[pid].turnaround_time = now - P[pid].arrival_time;
                }
                current_pid = -1;
            } else {
                /* still has CPU time left – re-queue */
                P[pid].state = STATE_READY;
                ready_push(pid);
                current_pid = -1;
            }

            int next = pick_next(P, algo);
            if (next != -1) {
                current_pid = next;
                P[current_pid].state = STATE_RUNNING;
                if (!P[current_pid].first_run) {
                    P[current_pid].first_run    = 1;
                    P[current_pid].response_time = now - P[current_pid].arrival_time;
                }
                busy_start = now;
                int burst = P[current_pid].remaining_cpu;
                int run   = (burst < time_quantum) ? burst : time_quantum;
                quantum_end = now + run;
                eq_push((Event){quantum_end, EVT_QUANTUM_EXPIRE, current_pid});
            }
            break;
        }

        case EVT_PREEMPT: break; /* handled inline above */
        }
    }

    /* ── Compute metrics ────────────────────────────────────────────────── */
    result->total_time = 0;
    double sum_wt = 0, sum_tat = 0, sum_rt = 0;
    int finished = 0;

    for (int i = 0; i < n; i++) {
        PCB *p = &P[i];
        if (p->state == STATE_DONE) {
            p->waiting_time = p->turnaround_time
                - p->bursts[0].cpu_burst; /* simplified: TAT - total CPU */
            /* better: TAT - sum of all CPU bursts */
            int total_cpu = 0;
            for (int b = 0; b < p->burst_count; b++)
                total_cpu += p->bursts[b].cpu_burst;
            p->waiting_time = p->turnaround_time - total_cpu;
            if (p->waiting_time < 0) p->waiting_time = 0;

            sum_wt  += p->waiting_time;
            sum_tat += p->turnaround_time;
            sum_rt  += (p->response_time >= 0) ? p->response_time : 0;
            finished++;
            if (p->finish_time > result->total_time)
                result->total_time = p->finish_time;
        }
    }

    if (finished > 0) {
        result->avg_waiting_time    = sum_wt  / finished;
        result->avg_turnaround_time = sum_tat / finished;
        result->avg_response_time   = sum_rt  / finished;
    }

    result->cpu_utilization = (result->total_time > 0)
        ? (double)total_busy / result->total_time * 100.0
        : 0.0;
    result->throughput = (result->total_time > 0)
        ? (double)finished / result->total_time * 1000.0  /* processes / sec if ms */
        : 0.0;
}
